precompiles/balances-erc20/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+					Runtime::Lookup::unlookup(to),
+					amount,
+				),
+				gasometer.remaining_gas()?,
+			)?;
+			gasometer.record_cost(used_gas)?;
+		}
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: vec![],
+			logs: LogsBuilder::new(context.address)
+				.log3(
+					SELECTOR_LOG_TRANSFER,
+					context.caller,
+					to,
+					EvmDataWriter::new().write(amount).build(),
+				)
+				.build(),
+		})
+	}
+
+	// This function is annotated with transactional.
+	// This is to ensure that if the substrate call fails, the change in allowance is reverted.
+	#[transactional]
+	fn transfer_from(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
+		gasometer.record_log_costs_manual(3, 32)?;
+
+		// Parse input.
+		input.expect_arguments(3)?;
+		let from: H160 = input.read::<Address>()?.into();
+		let to: H160 = input.read::<Address>()?.into();
+		let amount: U256 = input.read()?;
+
+		{
+			let caller: Runtime::AccountId =
+				Runtime::AddressMapping::into_account_id(context.caller);
+			let from: Runtime::AccountId = Runtime::AddressMapping::into_account_id(from.clone());
+			let to: Runtime::AccountId = Runtime::AddressMapping::into_account_id(to);
+			let amount = Self::u256_to_amount(amount)?;
+
+			// If caller is "from", it can spend as much as it wants.
+			if caller != from {
+				ApprovesStorage::<Runtime, Instance>::mutate(from.clone(), caller, |entry| {
+					// Get current value, exit if None.
+					let value = entry.ok_or(error("spender not allowed"))?;
+
+					// Remove "amount" from allowed, exit if underflow.
+					let new_value = value
+						.checked_sub(&amount)
+						.ok_or_else(|| error("trying to spend more than allowed"))?;
+
+					// Update value.
+					*entry = Some(new_value);
+
+					Ok(())
+				})?;
+			}
+
+			// Build call with origin. Here origin is the "from"/owner field.
+			// Dispatch call (if enough gas).
+			let used_gas = RuntimeHelper::<Runtime>::try_dispatch(
+				Some(from).into(),
+				pallet_balances::Call::<Runtime, Instance>::transfer(
+					Runtime::Lookup::unlookup(to),
+					amount,
+				),
+				gasometer.remaining_gas()?,
+			)?;
+			gasometer.record_cost(used_gas)?;
+		}
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: vec![],
+			logs: LogsBuilder::new(context.address)
+				.log3(
+					SELECTOR_LOG_TRANSFER,
+					from,
+					to,
+					EvmDataWriter::new().write(amount).build(),
+				)
+				.build(),
+		})
+	}
+
+	fn u256_to_amount(value: U256) -> EvmResult<BalanceOf<Runtime, Instance>> {
+		value
+			.try_into()
+			.map_err(|_| error("amount is too large for provided balance type"))
+	}
```

Comment:
Is this re-useable? Should it go in the utils?

Reply:
The issue I see is that it is using this crate `BalanceOf` type, and as we say each precompile crate pretty much have its own `BalanceOf`. That's also why I didn't add a read/write `BalanceOf` in the util.

Reply:
Here also a manual conversion would be more efficient. But it requires to break the genericity on the Balance and to assume that it is a u128.